### DISCUSSION TITLE
Controlled Slider and behaviour of onValueChange/onValueCommit

### DISCUSSION THREAD

**Discussion by stuartkeith** [radix-ui/primitives#2169] [Help]
Hi, I'm using the Slider component in a controlled way - the docs don't cover this use case (https://www.radix-ui.com/docs/primitives/components/slider)

The Radix Slider component has `onValueChange` and `onValueCommit`. So if I want the parent component to only update on the chosen value, I want to call `onChange` in `onValueCommit`, not `onValueChange`:

```
function SliderExample({ value, onChange ) {
  return <Slider
    value={[value]}
    // we want onChange to be called when the user has chosen a value, not while they drag the slider around
    onValueCommit={(values) => {
      onChange(values[0]);
    }}
}
```

But this doesn't work, the slider doesn't update at all.

It seems like I need to pass in the updated `value` from `onValueChange` myself. I can do this using some new state:

```
function SliderExample({ value, onChange ) {
  const [uncommittedValue, setUncommittedValue] = React.useState(null);

  return <Slider
    value={[uncommittedValue ?? value]}
    onValueChange={setUncommittedValue}
    onValueCommit={(values) => {
      setUncommittedValue(null);

      onChange(values[0]);
    }}
}
```

And that works fine for the pointer.

The problem I've encountered is keyboard events are the other way around - the `onValueCommit` is fired first, then `onValueChange` is fired. So `uncommittedValue` gets stuck in an invalid state - `onChange` is called and then `uncommittedValue` is set to a non-null value.

I'm not sure if this is a bug and the order should be the same between pointer/keyboard (the docs/tests don't mention how onValueChange/onValueCommit should be dealt with when value is controlled) or if I'm approaching it in the wrong way. Thanks for your time.

**Comment by benoitgrelard** [comment]
Hey @stuartkeith, that makes sense.

When you do this:


```jsx
function SliderExample({ value, onChange }) {
  return <Slider
    value={[value]}
    // we want onChange to be called when the user has chosen a value, not while they drag the slider around
    onValueCommit={(values) => {
      onChange(values[0]);
    }}
}
```

You are controlling the value of the component, setting it to the `value` you passed in, but that value is never updated as the thumb is being dragged.

There is no need to re-wrap the slider as you do, you can just use its API directly using the `onValueCommit` event to run some logic, or update some other state only when the value is commited.

**Reply by stuartkeith** [reply]
> There is no need to re-wrap the slider as you do, you can just use its API directly using the `onValueCommit` event to run some logic, or update some other state only when the value is commited.

Thanks for your reply. I'm not sure if I'm misunderstanding, but how is `value` handled in that case? This would work fine if I didn't need the slider to be controlled, but if I do, I'm not sure how not wrapping the slider stops this issue from coming up.

**Reply by benoitgrelard** [reply]
With any controlled prop, once you pass a value to it, your are telling the component what value should be displayed/used, the component then will tell you when to update that value, if you don't update it, it will never change.

For a native `input` that's basically `value` and `onChange` (where you'd do `setValue(event.target.value)` <= that's the component telling you what the new value is for your to update).

This is the same thing: `value` and `onValueChange`.

`onValueCommit` is just an extra event for you to hook into if you want to send value to a server (for example) only when the value has truly been set (rather than when the user is figuring out a value/dragging).

Hope that makes sense.

**Reply by stuartkeith** [reply]
Thanks. I think there was a misunderstanding - my `SliderExample` only wanted to update the parent state when the user commits to a value. That's why it used `onValueCommit`. I think there are use cases for both - sometimes the parent component needs to know about the user moving the pointer around, other times it only cares about the final committed value.

From your response it seems like it's assumed that the Slider will always want to update the state on `onValueChange`, as the user moves the pointer, and `onValueCommit` is more for side effects.

I think this is difficult to work around because the keyboard fires events in the opposite order to the pointer - the pointer is one or more onValueChanges followed by a onValueCommit, but the keyboard is a onValueCommit followed by an onValueChange. So using internal state for the temporary uncommitted value and clearing it on commit doesn't work (as `SliderExample` tried to do), and I can't think of another approach that doesn't involve starting to listen to keyboard/pointer events to try to override what is happening.

I've decided to go with an alternative component for my problem, but thanks for your help!
